_posts/2025-03-26-phased-strimzi-upgrade-example.md

```diff
@@ -0,0 +1,702 @@
+In this setup the roll back is a simple label change:
+
+```shell
+kubectl -n kafka label --overwrite kafka kafka-a strimzi-resource-selector=strimzi-0-43-0
+```
+
+```mermaid
+flowchart TD
+  S_043(Strimzi 0.43) --> K_A(Kafka A)
+  S_044(Strimzi 0.44) --> K_B(Kafka B)
+	S_044 --> K_C(Kafka C)
+```
+
+We fix the configuration issue and move Kafka cluster A over to the 0.44 operator.
+
+```mermaid
+flowchart TD
+  S_043(Strimzi 0.43)
+	style S_043 stroke:#f66,stroke-width:2px
+  S_044(Strimzi 0.44) --> K_A(Kafka A)
+  S_044 --> K_B(Kafka B)
+	S_044 --> K_C(Kafka C)
+```
+
+At this point the Strimzi 0.43 operator is not managing any Kafka clusters and all the clusters being managed by the 0.44 operator have been stable for a sufficient period that we are confident we don't need to roll back. 
+We can now remove the Strimzi 0.43 operator deployment and its associated RBAC resources.
+At this point we can add a new version to the Operator Set, adding the 0.45 installation files and upgrading the CRDs to those of the 0.45 version.
+
+```mermaid
+flowchart TD
+  S_044(Strimzi 0.44) --> K_A(Kafka A)
+  S_044 --> K_B(Kafka B)
+	S_044 --> K_C(Kafka C)
+  S_045(Strimzi 0.45)
+	style S_045 stroke:#008000,stroke-width:2px
+```
+
+#### Things to watch out for
+
+The main things to watch for when using this Strimzi deployment pattern are:
+- Take note of the advice in the [CRD section](#custom-resource-definitions-1) around the number of concurrent versions and CRD API version upgrades:
+  - The only supported configuration is running two adjacent Strimzi versions concurrently.
+  - When upgrading the CRD API version, you must first upgrade all operands to the last Strimzi version that uses the old API, before adding the new Strimzi version with the updated CRDs.
+- Take extreme care when updating the Operator Set to avoid removing the CRDs from the Kubernetes cluster. Doing so would delete all Kafka clusters and other Strimzi-managed resources.
+- You should ensure that the Kafka version of the Kafka CR is supported by the version of Strimzi you are moving too. This should not be an issue if you are moving between adjacent Strimzi versions. However, you can add checks to deployment scripts and GitOps repository CI.
+- When removing old versions of Strimzi that are no longer needed, care should be taken to make sure that no Kafka CRs are still labelled for that version. Again, scripting or CI can run checks against the Kubernetes cluster to ensure this.
+
+#### Further deployment options
+
+We have covered how to do upgrades using the custom resource selector and multiple different versions of Strimzi. 
+However, there is no reason there should be only one instance of each version.
+You could have `strimzi-0-43-0-a` and `stimzi-0-43-0-b`, you could then load balance Kafka cluster management between them.
```

Comment:
I think _load balancing_ suggests it happens automatically. Would sharding be better here?

Reply:
Good point, I went with "distribute between them".